### PR TITLE
Toolchain part2: `./mkmap_data.py`

### PR DESCRIPTION
**Pull request by negin513**
### Description of changes

This Python script is the second part of the simplified toolchain for creating surface datasets for global ctsm cases (#644 ).

After creating the namelist/control file with `./gen_mksurf_namelist.py` (PR #1419) using the options for your desired case, you should run :
```
./mkmap_data.py --namelist [namelist from ./gen_mksurf_namelist.py]
```
to create mapping/weight files.

This python script is the alternative code to mkmapdata.sh with few differences:

1. It reads namelist file for DST mesh file.
                
2. For raw dataset in the namelist it finds their SRC mesh file and mask from
the necdf metadata.
                
3. Check if the weight (mapping file) already exists or not. 
                
4. If it does not exist, it creates the mapping (weight) file. 

### Specific notes

Contributors other than yourself, if any: @slevisconsulting 

CTSM Issues Fixed (include github issue #):

Are answers expected to change (and if so in what way)?

Any User Interface Changes (namelist or namelist defaults changes)?

`./mkmap_data.py` has the following interface:
```
usage: mkmap_data.py [-h] --namelist NAMELIST [--input_dir INPUT_DIR]
                     [--map_dir MAP_DIR] [-d] [--overwrite]
                     [--esmf_path OVERWRITE] [--account ACCOUNT_NAME]

|------------------------------------------------------------------|
|---------------------  Instructions  -----------------------------|
|------------------------------------------------------------------|
This Python script is part of the simplified toolchain for creating
the surface dataset for ctsm cases.

After creating the namelist/control file with ./gen_mksurf_namelist.py
using the options for your desired case, you should run :

./mkmap_data.py --namelist [namelist from ./gen_mksurf_namelist.py]

to create mapping files.

This python script is the alternative code to mkmapdata.sh with
 few differences:

1. It reads namelist file for DST mesh file.

2. For raw dataset in the namelist it finds their SRC mesh file and mask from
the necdf metadata.

3. Check if the weight (mapping file) already exists or not. 

4. If it does not exist, it creates the mapping (weight) file. 
 
-------------------------------------------------------------------
Instructions for running on Cheyenne/Casper:
 
load the following into your local environment:
 
    module load python
    ncar_pylib
-------------------------------------------------------------------
To see the available options:
    ./mkmap_data.py --help
 
To run the script:
    ./mkmap_data.py --namelist ${namelist from ./gen_mksurf_namelist.py}
 
To remove NPL from your environment on Cheyenne/Casper:
    deactivate
-------------------------------------------------------------------

optional arguments:
  -h, --help            show this help message and exit
  --namelist NAMELIST   namelist for the ctsm case created by
                        gen_mksurf_namelist.py
  --input_dir INPUT_DIR
                        Path of your mesh files and input data.', [default:
                        /glade/p/cesm/cseg/inputdata/]
  --map_dir MAP_DIR     Path of your weight (mapping) files.', [default: /glad
                        e/scratch/negins/ctsm_toolchain_2/tools/toolchain]
  -d, --debug           Debug mode will print more information [default:
                        False]
  --overwrite           Flag to overwrite the existing mapping files.
                        [default: False]
  --esmf_path OVERWRITE
                        Path of the ESMF library. [default: None]
  --account ACCOUNT_NAME
                        Cheyenne or Casper project account name.', [default:
                        P93300606]

```
Testing performed, if any: @slevisconsulting

### CODE REVIEW

File: tools/toolchain/mkmap_data.py [hunk 1]
```diff
@@ -0,0 +1,798 @@
...
+    dst_mesh : MeshType
+        Destination mesh in MeshType which contains the information of
+        dst mesh file.
+
+    map_dir : str
+        Directory where you want to store your weight/mapping files.
+
+    mapping_fname: str
+        Mapping /weight filename, if not explicitly given the
+        class will create one based on src_mesh and dst_mesh information.
```
Comment (slevisconsulting) [inline tools/toolchain/mkmap_data.py:433]: This is similar to res and mask, where the advanced user may choose to supply their own value, right? I have the same concern here as I do about res and mask.

**Review by slevisconsulting — COMMENTED** [review]
@negin513 thank you for your nice work on the toolchain! Also, thank you for giving me a tour of the PR this morning.

I have a very important general comment that I realized after we spoke:
According to the design that the group adopted [(slide 5 in this presentation)](https://docs.google.com/presentation/d/1jYYbJtulDfWDV_CToGAweIPT-S36gxpesjQeLCmMKzA/edit?usp=sharing), this piece of code should also generate the "old" namelist (the one that the fortran code reads). The "old" namelist has what the new namelist contains PLUS the list of map_ files. I think that's all.

File: tools/toolchain/mkmap_data.py [hunk 2]
```diff
@@ -0,0 +1,798 @@
...
+import logging
+import argparse
+import subprocess
+
+import xarray as xr
+
+from datetime import datetime
+from getpass import getuser
+
+# from gen_mksurf_namelist import get_parser, build_nl
```
[the remark this answers is not in the document]
Comment (slevisconsulting) [reply tools/toolchain/mkmap_data.py:92]: is this a placeholder to address the generation of the "old" namelist?

**Comment by slevisconsulting** [comment]
> According to the design that the group adopted [(slide 5 in this presentation)](https://docs.google.com/presentation/d/1jYYbJtulDfWDV_CToGAweIPT-S36gxpesjQeLCmMKzA/edit?usp=sharing), this piece of code should also generate the "old" namelist (the one that the fortran code reads). The "old" namelist has what the new namelist contains PLUS the list of map_ files. I think that's all.

@negin513 recommended that it would be cleaner to put the function of creating the old namelist in a separate script that we call gen_old_namelist.py. See updated slides (same link). Also gen_old_namelist.py will be a step that's transparent to the user and will be run by the wrapper script. The wrapper script will be a new pull request (part 3 in this series of PRs).

**Comment by ekluzek** [comment]
With the work on the parallel mksurfdata_map, this PR will likely be closed without coming in.

**Comment by ekluzek** [comment]
With the work in #1663 this work becomes obsolete so I'm closing. Some of the python code in here might still be useful examples that we keep in mind. One thing in here is one of our examples of interacting with the batch system.

**Pull request closed without merging**